app/kubemci/pkg/ingress/ingresssyncer.go

```diff
@@ -0,0 +1,152 @@
+package ingress
+
+import (
+	"fmt"
+	"io/ioutil"
+	"reflect"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	multierror "github.com/hashicorp/go-multierror"
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/diff"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/ingress-gce/pkg/annotations"
+)
+
+var (
+	defaultIngressNamespace = "default"
+)
+
+type IngressSyncer struct {
+}
+
+func NewIngressSyncer() IngressSyncerInterface {
+	return &IngressSyncer{}
+}
+
+func (i *IngressSyncer) EnsureIngress(ing *v1beta1.Ingress, clients map[string]kubeclient.Interface, forceUpdate bool) ([]string, error) {
+	var err error
+	var ensuredClusters []string
+	for cluster, client := range clients {
+		glog.V(4).Infof("Creating Ingress in cluster: %v...", cluster)
+		ensuredClusters = append(ensuredClusters, cluster)
+		glog.V(3).Infof("Using namespace %s for ingress %s", ing.Namespace, ing.Name)
+		existingIng, getErr := client.ExtensionsV1beta1().Ingresses(ing.Namespace).Get(ing.Name, metav1.GetOptions{})
+		if getErr != nil {
+			if errors.IsNotFound(getErr) {
+				actualIng, createErr := client.ExtensionsV1beta1().Ingresses(ing.Namespace).Create(ing)
+				glog.V(2).Infof("Ingress Create returned: err:%v. Actual Ingress:%+v\n", createErr, actualIng)
+				if createErr != nil {
+					err = multierror.Append(err, fmt.Errorf("Error in creating ingress in cluster %s: %s", cluster, createErr))
+				}
+			} else {
+				err = multierror.Append(err, fmt.Errorf("Error in checking for existing ingress in cluster %s: %s", cluster, getErr))
+			}
+			fmt.Println("Created Ingress in cluster:", cluster)
+		} else {
+			if !reflect.DeepEqual(ing.Spec, existingIng.Spec) {
```

Comment:
Umm we should compare the whole object and not just Spec? Specially ObjectMeta for when user wants to add an annotation?

Reply:
Yes definitely!